js/viz/sankey/layout.js

```diff
@@ -0,0 +1,265 @@
+    _weightPerPixel: null,
+
+    _getCascadeIdx: function(nodeTitle, cascadesConfig) {
+        let nodeInfo = cascadesConfig.find((c) => { return c.name === nodeTitle; });
+
+        if(nodeInfo.outgoing.length > 0) {
+            // in common case number of cascade is the longest path to the node
+            return nodeInfo.lp;
+        } else {
+            // but if the node is last (ie has no outgoing edges) lets move it to the LAST cascade
+            return graphModule.routines.maxOfArray(cascadesConfig.map(c => { return c.lp; }));
+        }
+    },
+
+    _getInWeightForNode: function(nodeTitle, links) {
+        let w = 0;
+        links.forEach((link) => {
+            if(link[1] === nodeTitle) {
+                w += link[2];
+            }
+        });
+        return w;
+    },
+
+    _getOutWeightForNode: function(nodeTitle, links) {
+        let w = 0;
+        links.forEach((link) => {
+            if(link[0] === nodeTitle) {
+                w += link[2];
+            }
+        });
+        return w;
+    },
+
+    _computeCascades: function(links) {
+        let cascadesConfig = graphModule.struct.computeLongestPaths(links),
+            maxCascade = graphModule.routines.maxOfArray(cascadesConfig.map(c => { return c.lp; }));
+
+        let cascades = [];
+
+        // init cascades
+        for(let i = 0; i < maxCascade + 1; i++) {
+            cascades.push({});
+        }
+
+        // add nodes to cascades
+        links.forEach((link) => {
+            let cascade = cascades[this._getCascadeIdx(link[0], cascadesConfig)];
+            if(!cascade[link[0]]) {
+                cascade[link[0]] = { nodeTitle: link[0] };
+            }
+            cascade = cascades[this._getCascadeIdx(link[1], cascadesConfig)];
+            if(!cascade[link[1]]) {
+                cascade[link[1]] = { nodeTitle: link[1] };
+            }
+        });
+
+        // compute in and out weightes of all nodes of cascades
+        cascades.forEach(cascade => {
+            Object.keys(cascade).forEach(nodeTitle => {
+                let node = cascade[nodeTitle];
+                node.inWeight = this._getInWeightForNode(node.nodeTitle, links);
+                node.outWeight = this._getOutWeightForNode(node.nodeTitle, links);
+                node.maxWeight = Math.max(node.inWeight, node.outWeight);
+            });
+        });
+
+        return cascades;
+    },
+
+    _getWeightForCascade: function(cascades, cascadeIdx) {
+        let wIn = 0, wOut = 0, wMax = 0, cascade = cascades[cascadeIdx];
+        Object.keys(cascade).forEach(nodeTitle => {
+            wIn += cascade[nodeTitle].inWeight;
+            wOut += cascade[nodeTitle].outWeight;
+            wMax += Math.max(cascade[nodeTitle].inWeight, cascade[nodeTitle].outWeight);
+        });
+        return wMax;
+    },
+
+    _getMaxWeightThroughCascades: function(cascades) {
+        let max = [];
+        cascades.forEach(cascade => {
+            let inW = 0, outW = 0, mW = 0;
```

Comment:
Some of these variables were not used. I have removed them [here](https://github.com/DevExpress/DevExtreme/pull/4909/commits/39cb78c426f6c861e249a689e9eab014287a08c2#diff-66fa2bb67e5bf1b7dbe5e0584c55a4deR91). Maybe this method needs additional refactoring now.